File: problems/027/027.c
#include <stdio.h>
#include <stdlib.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    FILE *fp;
    char prime_num[10];
    const char* FILE_NAME = "problems/003/PrimeNumbers_Upto_1000000";
    fp = fopen(FILE_NAME, "r");

    int *primes_list = (int*)calloc(1E3, sizeof(int));

    int a       = -999;
    int b_count =    0;

    int a_max = 0;
    int b_max = 0;
    int n_max = 0;

    int i = 0;
    while ((fgets(prime_num, 10, fp) != NULL) && (i != 1000))
    {
        if (atoi(prime_num) < 1000)
            b_count++;

        primes_list[i++] = atoi(prime_num);
    }

    while (a < 1000)
    {
        int b = 0;
        while (b < b_count)
        {
            int n = 1;
            int j = 0;
            int max_prime = (n * n) + (a * n) + primes_list[b];

            while (j < 1E3)
            {
                if (primes_list[j] > max_prime)
                    break;

                if (max_prime == primes_list[j])
                {
                    n++;
                    j = 0;
                    max_prime = (n * n) + (a * n) + primes_list[b];
                    continue;
                }

                j++;
            }

            if (n > n_max)
            {
                n_max = n;
                a_max = a;
                b_max = primes_list[b];
            }

            b++;
        }

        a++;
    }

    fclose(fp);
    free(primes_list);

    printf("%d\n", a_max * b_max);

    return 0;
}

File: problems/027/027.f90
PROGRAM F027

    ! Copyright 2021 Melwyn Francis Carlo
    ! FILE REFERENCE: http://www.naturalnumbers.org/primes.html

    IMPLICIT NONE

    INTEGER :: PRIME_NUM
    CHARACTER (LEN=50), PARAMETER :: FILENAME = "problems/003/PrimeNumbers_Upto_1000000"

    INTEGER, DIMENSION(1000) :: PRIMES_LIST

    INTEGER :: A       = -999
    INTEGER :: B_COUNT =    0

    INTEGER :: A_MAX   = 0
    INTEGER :: B_MAX   = 0
    INTEGER :: N_MAX   = 0

    INTEGER :: I = 1

    INTEGER :: B, N, J, MAX_PRIME

    OPEN (2, FILE = FILENAME, STATUS = 'OLD')

    DO WHILE (I /= 1000)
        READ(2, *, END=10) PRIME_NUM
        IF (PRIME_NUM < 1000) B_COUNT = B_COUNT + 1
        PRIMES_LIST(I) = PRIME_NUM
        I = I + 1
    END DO

    10 CLOSE(2)

    DO WHILE (A < 1000)

        B = 1
        DO WHILE (B < B_COUNT)

            N = 1
            J = 1
            MAX_PRIME = (N * N) + (A * N) + PRIMES_LIST(B)

            DO WHILE (J < 1000)

                IF (PRIMES_LIST(J) > MAX_PRIME) EXIT

                IF (MAX_PRIME == PRIMES_LIST(J)) THEN
                    N = N + 1
                    J = 1
                    MAX_PRIME = (N * N) + (A * N) + PRIMES_LIST(B)
                    CYCLE
                END IF

                J = J + 1

            END DO

            IF (N > N_MAX) THEN
                N_MAX = N
                A_MAX = A
                B_MAX = PRIMES_LIST(B)
            END IF

            B = B + 1

        END DO

        A = A + 1

    END DO

    PRINT ('(I0)'), A_MAX * B_MAX

END PROGRAM F027
